src/states/adult/mod.rs

```diff
@@ -539,63 +459,127 @@ impl Base for Adult {
-                self.handle_bootstrap_request(p2p_node, name);
-                Ok(Transition::Stay)
+            Variant::ParsecRequest(version, request) => self.handle_parsec_request(
+                version,
+                request,
+                msg.src.to_sender_node(sender)?,
+                outbox,
+            ),
+            Variant::ParsecResponse(version, response) => {
+                self.handle_parsec_response(version, response, *msg.src.as_node()?, outbox)
             }
-            ConnectionResponse => {
-                debug!("{} - Received connection response from {}", self, p2p_node);
+            Variant::BootstrapRequest(name) => {
+                self.handle_bootstrap_request(msg.src.to_sender_node(sender)?, name);
                 Ok(Transition::Stay)
             }
-            JoinRequest(_) | MemberKnowledge { .. } => {
+            Variant::ConnectionResponse => {
                 debug!(
-                    "{} Unhandled direct message from {}, adding to backlog: {:?}",
+                    "{} - Received connection response from {}",
                     self,
-                    p2p_node.public_id(),
-                    msg
+                    msg.src.to_sender_node(sender)?
                 );
-                self.msg_backlog.push(QueuedMessage::Direct(p2p_node, msg));
                 Ok(Transition::Stay)
             }
-            NeighbourInfo(_)
-            | UserMessage(_)
-            | NodeApproval(_)
-            | AckMessage { .. }
-            | GenesisUpdate(_)
-            | Relocate(_)
-            | BootstrapResponse(_) => {
-                debug!(
-                    "{} Unhandled direct message from {}, discard: {:?}",
-                    self,
-                    p2p_node.public_id(),
-                    msg
-                );
+            Variant::NeighbourInfo(_)
+            | Variant::UserMessage(_)
+            | Variant::NodeApproval(_)
+            | Variant::AckMessage { .. }
+            | Variant::JoinRequest(_)
+            | Variant::MemberKnowledge(_) => {
+                debug!("{} Unhandled message, adding to backlog: {:?}", self, msg);
+                self.msg_backlog.push(msg.into_queued(sender));
+                Ok(Transition::Stay)
+            }
+            Variant::BootstrapResponse(_) => {
+                debug!("{} Unhandled message, discarding: {:?}", self, msg);
                 Ok(Transition::Stay)
             }
         }
     }
 
-    fn handle_hop_message(
-        &mut self,
-        msg: HopMessageWithBytes,
-        _outbox: &mut dyn EventBox,
-    ) -> Result<Transition, RoutingError> {
-        self.handle_signed_message(msg)
+    fn filter_incoming_message(&mut self, message: &MessageWithBytes) -> bool {
+        self.msg_filter.filter_incoming(message).is_new()
+    }
+
+    fn should_verify_message(&self, msg: &Message) -> bool {
+        // As Adult, we only verify the messages that we know how to process ourselves.
```

Comment:
Better but still not ideal as is, I'd be happier if we can make that clearer.
Would it make sense to split the function and do something like that?

```rust
         if let Some(msg) = self.preprocess_message(msg)? {
            if self.should_verify_message(msg) {
                 self.verify_message(&msg)?;
                 self.handle_message(None, msg, outbox)
            } else {
                 self.handle_unverified_message(&msg)
            }
        } else {
            Ok(Transition::Stay)
        } 
````

Reply:
Initial comment was https://github.com/maidsafe/routing/pull/2030#discussion_r372307058

Reply:
Sounds reasonable.